coderd/workspaceagents_test.go

```diff
@@ -2641,3 +2644,91 @@ func TestAgentConnectionInfo(t *testing.T) {
 	require.True(t, info.DisableDirectConnections)
 	require.True(t, info.DERPForceWebSockets)
 }
+
+func TestReinit(t *testing.T) {
+	t.Parallel()
+
+	t.Run("unclaimed workspaces are not reinitialized", func(t *testing.T) {
+		t.Parallel()
+
+		if !dbtestutil.WillUsePostgres() {
+			t.Skip()
+		}
+
+		client, db := coderdtest.NewWithDatabase(t, nil)
+		user := coderdtest.CreateFirstUser(t, client)
+
+		r := dbfake.WorkspaceBuild(t, db, database.WorkspaceTable{
+			OrganizationID: user.OrganizationID,
+			OwnerID:        user.UserID,
+		}).WithAgent().Do()
+		ctx := testutil.Context(t, testutil.WaitShort)
+		err := db.UpdateProvisionerJobWithCompleteByID(ctx, database.UpdateProvisionerJobWithCompleteByIDParams{
+			ID:        r.Build.JobID,
+			UpdatedAt: time.Now(),
+			CompletedAt: sql.NullTime{
+				Valid: true,
+				Time:  time.Now(),
+			},
+		})
+		require.NoError(t, err)
+
+		agentCtx := testutil.Context(t, testutil.WaitShort)
+		agentClient := agentsdk.New(client.URL)
+		agentClient.SetSessionToken(r.AgentToken)
+		reinitEvent, err := agentClient.WaitForReinit(agentCtx)
+		require.ErrorIs(t, err, context.DeadlineExceeded)
+		require.Nil(t, reinitEvent)
+	})
+	t.Run("claimed workspaces are reinitialized", func(t *testing.T) {
+		t.Parallel()
+
+		if !dbtestutil.WillUsePostgres() {
+			t.Skip()
+		}
+
+		db, ps := dbtestutil.NewDB(t)
+		client := coderdtest.New(t, &coderdtest.Options{
+			Database: db,
+			Pubsub:   ps,
+		})
+		user := coderdtest.CreateFirstUser(t, client)
+
+		r := dbfake.WorkspaceBuild(t, db, database.WorkspaceTable{
+			OrganizationID: user.OrganizationID,
+			OwnerID:        user.UserID,
+		}).WithAgent().Do()
+		ctx := testutil.Context(t, testutil.WaitShort)
+		err := db.UpdateProvisionerJobWithCompleteByID(ctx, database.UpdateProvisionerJobWithCompleteByIDParams{
+			ID:        r.Build.JobID,
+			UpdatedAt: time.Now(),
+			CompletedAt: sql.NullTime{
+				Valid: true,
+				Time:  time.Now(),
+			},
+		})
+		require.NoError(t, err)
+
+		agentCtx := testutil.Context(t, testutil.WaitShort)
+		agentClient := agentsdk.New(client.URL)
+		agentClient.SetSessionToken(r.AgentToken)
+
+		agentReinitializedCh := make(chan *agentsdk.ReinitializationEvent)
+		go func() {
+			reinitEvent, err := agentClient.WaitForReinit(agentCtx)
+			assert.NoError(t, err)
+			agentReinitializedCh <- reinitEvent
+		}()
+
+		time.Sleep(time.Second)
```

Comment:
sleeping in unit tests is a code smell. It means you're not actually writing a deterministic test.

The race here seems to be that you don't know that the `WaitForReinit` call has actually subscribed to the pubsub before you can send your claim.

One possibility is to fake/mock the pubsub, so that you can capture the subscription. The disadvantage is that you're passing the pubsub to the whole coderd API, so there may be a lot of spurious calls to deal with.

The other option is to make the WorkspaceClaimListener an interface, and mock it out. 

In either case, to make this easier without having to pass things through the entire Coderd API, you could separate out the handler into a struct with just the stuff it needs (database, logger, claim listener) , and then test that.

Reply:
fixed.